src/broadcast/mpris/mprisplayer.cpp

```diff
@@ -0,0 +1,457 @@
+    CoverInfo coverInfo = pTrack->getCoverInfoWithLocation();
+    if (coverInfo.type == CoverInfoRelative::FILE) {
+        QFileInfo fileInfo;
+        if (!coverInfo.trackLocation.isEmpty()) {
+            fileInfo = QFileInfo(coverInfo.trackLocation);
+        }
+        QFileInfo coverFile(fileInfo.dir(), coverInfo.coverLocation);
+        QString path = coverFile.absoluteFilePath();
+        qDebug() << "Cover art path: " << path;
+        QUrl fileUrl = QUrl::fromLocalFile(path);
+        if (!fileUrl.isValid()) {
+            qDebug() << "Invalid URL: " << fileUrl;
+            return;
+        }
+        m_currentMetadata.coverartUrl = fileUrl.toString();
+    } else if (coverInfo.type == CoverInfoRelative::METADATA) {
+        CoverArtCache::requestCover(this, coverInfo);
+    } else {
+        m_currentMetadata.coverartUrl.clear();
+    }
+}
+
+void MprisPlayer::slotPlayChanged(DeckAttributes* pDeck, bool playing) {
+    Q_UNUSED(pDeck);
+    Q_UNUSED(playing);
+    m_pMpris->notifyPropertyChanged(kPlayerInterfaceName,
+            "PlaybackStatus",
+            playbackStatus());
+}
+
+void MprisPlayer::slotPlayPositionChanged(DeckAttributes* pDeck, double position) {
+    if (autoDjIdle()) {
+        qlonglong playPosition = static_cast<qlonglong>(position * //Fraction of duration
+                pDeck->getLoadedTrack()->getDuration() *           //Duration in seconds
+                1e6);
+        m_pMpris->notifyPropertyChanged(kPlayerInterfaceName, "Position", playPosition);
+    }
+}
+
+DeckAttributes* MprisPlayer::findPlayingDeck() const {
+    for (int i = 0; i < m_deckAttributes.count(); ++i) {
+        if (m_deckAttributes.at(i)->isPlaying()) {
+            return m_deckAttributes.at(i);
+        }
+    }
+    return nullptr;
+}
+
+void MprisPlayer::slotMasterGainChanged(double value) {
+    m_pMpris->notifyPropertyChanged(kPlayerInterfaceName, "Volume", value);
+}
+
+double MprisPlayer::rate() const {
+    DeckAttributes* playingDeck = findPlayingDeck();
+    if (playingDeck != nullptr) {
+        ControlProxy rate(ConfigKey(
+                PlayerManager::groupForDeck(playingDeck->index - 1), "rate"));
+        return rate.get();
+    }
+    return 0;
+}
+
+void MprisPlayer::setRate(double value) {
+    DeckAttributes* playingDeck = findPlayingDeck();
+    if (playingDeck != nullptr) {
+        ControlProxy rate(ConfigKey(
+                PlayerManager::groupForDeck(playingDeck->index - 1), "rate"));
+        double clampedValue = qBound(-1.0, value, 1.0);
+        rate.set(clampedValue);
+    }
+}
+
+void MprisPlayer::slotCoverArtFound(const QObject* requestor,
+        const CoverInfoRelative& info,
+        const QPixmap& pixmap,
+        bool fromCache) {
+    Q_UNUSED(info);
+    Q_UNUSED(fromCache);
+
+    if (!pixmap.isNull() && requestor == this) {
+        QImage coverImage = pixmap.toImage();
+        m_currentCoverArtFile.open();
+        bool success = coverImage.save(&m_currentCoverArtFile, "JPG");
```

Comment:
Could we use PNG to avoid lossy compressing the cover art (which may already be a lossy compressed JPG)?